Samples/2.3/FasterSample/src/Grains/FasterThreadPoolGrain.cs

```diff
@@ -0,0 +1,408 @@
+                        throw new ApplicationException();
+                    }
+                }
+            }
+            finally
+            {
+                try
+                {
+                    if (session != Guid.Empty) lookup.StopSession();
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+        });
+
+        public Task SnapshotAsync() => Task.Run(async () =>
+        {
+            await semaphore.WaitAsync();
+
+            var session = Guid.Empty;
+            try
+            {
+                session = lookup.StartSession();
+                if (!lookup.CompletePending(true)) throw new ApplicationException();
+                if (!lookup.TakeFullCheckpoint(out var token)) throw new ApplicationException();
+                if (!lookup.CompleteCheckpoint(true)) throw new ApplicationException();
+            }
+            finally
+            {
+                try
+                {
+                    if (session != Guid.Empty) lookup.StopSession();
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+        });
+
+        public Task<LookupItem> TryGetAsync(int key) => Task.Run(async () =>
+        {
+            await semaphore.WaitAsync();
+
+            var session = Guid.Empty;
+            try
+            {
+                session = lookup.StartSession();
+                LookupItem result = null;
+                if (lookup.Read(ref key, ref result, ref result, Empty.Default, 0) == Status.ERROR)
+                {
+                    throw new ApplicationException();
+                }
+                return result;
+            }
+            finally
+            {
+                try
+                {
+                    if (session != Guid.Empty) lookup.StopSession();
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+        });
+
+        public Task<ImmutableList<LookupItem>> TryGetRangeAsync(ImmutableList<int> keys) => Task.Run(async () =>
+        {
+            await semaphore.WaitAsync();
+
+            var session = Guid.Empty;
+            try
+            {
+                session = lookup.StartSession();
+                var result = ImmutableList.CreateBuilder<LookupItem>();
+                foreach (var key in keys)
+                {
+                    var _key = key;
+                    LookupItem input = null;
+                    LookupItem output = null;
+                    switch (lookup.Read(ref _key, ref input, ref output, Empty.Default, 0))
+                    {
+                        case Status.OK:
+                            result.Add(output);
+                            break;
+
+                        case Status.PENDING:
+                            if (lookup.CompletePending(true))
```

Comment:
What you have is correct. However, you don't have to call CompletePending immediately after every operation in the batch. This is because CompletePending(true) spin-waits until all pending requests until then are completed. Thus all operations become synchronous.

Instead, you can simply call CompletePending(true) once after foreach, in order to complete all pending operations until that point.

Note that the pending operation results are reported via callback code (such as Functions.ReadCompletionCallback) during CompletePending(). You would therefore need to store the results array and the correct offset in the results array inside Context, so that ReadCompletionCallback can populate the result at the correct position in the result array.

Reply:
This grain now uses a `TaskCompletionSource<LookupItem>` as the operation context. Either the grain or the callback will resolve that completion, depending on whether the operation completed synchronously or asynchronously. The trade-off here is more allocations. I could over-engineer this with object pools and value tasks but I think there's more merit in changing the underlying operations so they are async themselves as per https://github.com/microsoft/FASTER/pull/130, so I'll look into that instead.